test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h

```diff
@@ -48,62 +57,196 @@ class FuzzingEventEngine : public EventEngine {
+    bool Write(SliceBuffer* data, int index) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  };
+
+  // Implementation of Endpoint.
+  // When a connection is formed, we create two of these - one with index 0, the
+  // other index 1, both pointing to the same EndpointMiddle.
+  class FuzzingEndpoint final : public Endpoint {
+   public:
+    FuzzingEndpoint(std::shared_ptr<EndpointMiddle> middle, int index)
+        : middle_(std::move(middle)), index_(index) {}
+    ~FuzzingEndpoint() override;
+
+    bool Read(absl::AnyInvocable<void(absl::Status)> on_read,
+              SliceBuffer* buffer, const ReadArgs* args) override;
+    bool Write(absl::AnyInvocable<void(absl::Status)> on_writable,
+               SliceBuffer* data, const WriteArgs* args) override;
+    const ResolvedAddress& GetPeerAddress() const override {
+      return middle_->addrs[peer_index()];
+    }
+    const ResolvedAddress& GetLocalAddress() const override {
+      return middle_->addrs[my_index()];
+    }
+
+   private:
+    int my_index() const { return index_; }
+    int peer_index() const { return 1 - index_; }
+    // Schedule additional writes to be performed later.
+    // Takes a ref to middle instead of holding this, so that should the
+    // endpoint be destroyed we don't have to worry about use-after-free.
+    // Instead that scheduled callback will see the middle is closed and finally
+    // report completion to the caller.
+    // Since there is no timeliness contract for the completion of writes after
+    // endpoint shutdown, it's believed this is a legal implementation.
+    static void ScheduleDelayedWrite(
+        std::shared_ptr<EndpointMiddle> middle, int index,
+        absl::AnyInvocable<void(absl::Status)> on_writable, SliceBuffer* data)
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+    std::shared_ptr<EndpointMiddle> middle_;
+    int index_;
+  };
+
+  void RunLocked(absl::AnyInvocable<void()> closure)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    RunAfterLocked(Duration::zero(), std::move(closure));
+  }
+
+  TaskHandle RunAfterLocked(Duration when, absl::AnyInvocable<void()> closure)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+  // Allocate a port. Considered fuzzer selected port orderings first, and then
+  // falls back to an exhaustive incremental search from port #1.
+  int AllocatePort() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // Is the given port in use by any listener?
+  bool IsPortUsed(int port) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  // For the next connection being built, query the list of fuzzer selected
+  // write size limits.
+  std::queue<size_t> WriteSizesForConnection()
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
   gpr_timespec NowAsTimespec(gpr_clock_type clock_type)
       ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
   static gpr_timespec GlobalNowImpl(gpr_clock_type clock_type)
       ABSL_LOCKS_EXCLUDED(mu_);
   const Duration final_tick_length_;
 
-  grpc_core::Mutex mu_;
+  static grpc_core::NoDestruct<grpc_core::Mutex> mu_;
```

Comment:
This should come with a big warning at the top that says there can only be exactly one FuzzingEventEngine alive at any given time. I know that's how it's intended to be used today, but I this class might get used more widely.

Reply:
added a comment: I think we need the constraint.

The only other viable design moves the singleton parts out and has multiple FuzzingEventEngine's reaching in - but I'm not sure that that's better.

Reply:
The current design seems good for now, we can revisit the decision later if we need.